charts/invenio/templates/_helpers.tpl

```diff
@@ -199,79 +199,110 @@ app.kubernetes.io/managed-by: {{ .Release.Service }}
   {{- if .Values.postgresql.enabled -}}
-    {{- required "Missing .Values.postgresql.primary.service.ports.postgresql" .Values.postgresql.primary.service.ports.postgresql -}}
+    {{- required "Missing .Values.postgresql.primary.service.ports.postgresql" (tpl .Values.postgresql.primary.service.ports.postgresql .) -}}
   {{- else -}}
-    {{- required "Missing .Values.postgresqlExternal.port" .Values.postgresqlExternal.port -}}
+    {{- required "Missing .Values.postgresqlExternal.port" (tpl .Values.postgresqlExternal.port .) -}}
   {{- end -}}
 {{- end -}}
 
-############################     Database name     ############################
 {{/*
   This template renders the name of the database in PostgreSQL.
 */}}
-{{- define "invenio.postgresql.databaseName" -}}
+{{- define "invenio.postgresql.database" -}}
   {{- if .Values.postgresql.enabled -}}
-    {{- required "Missing .Values.postgresql.auth.database" .Values.postgresql.auth.database -}}
+    {{- required "Missing .Values.postgresql.auth.database" (tpl .Values.postgresql.auth.database .) -}}
   {{- else -}}
-    {{- required "Missing .Values.postgresqlExternal.databaseName" .Values.postgresqlExternal.databaseName -}}
+    {{- required "Missing .Values.postgresqlExternal.database" (tpl .Values.postgresqlExternal.database .) -}}
   {{- end -}}
 {{- end -}}
 
-#######################     SQLAlchemy database URI     #######################
 {{/*
-  This template renders the SQLAlchemy database URI.
+  Define database connection env section. 
 */}}
-{{- define "invenio.sqlAlchemyDbUri" -}}
-  {{- $username := include "invenio.postgresql.username" . -}}
-  {{- $password := include "invenio.postgresql.password" . -}}
-  {{- $hostname := include "invenio.postgresql.hostname" . -}}
-  {{- $port := include "invenio.postgresql.port" . -}}
-  {{- $databaseName := include "invenio.postgresql.databaseName" . -}}
-  {{- printf "postgresql+psycopg2://%s:%s@%s:%v/%s" $username $password $hostname $port $databaseName -}}
+{{- define "invenio.config.database" -}}
+- name: INVENIO_DB_USER
+  value: {{ include "invenio.postgresql.username" . }}
+- name: INVENIO_DB_HOST
+  value: {{ include "invenio.postgresql.hostname" . }}
+- name: INVENIO_DB_PORT
+  value: {{ include "invenio.postgresql.port" . }}
+- name: INVENIO_DB_NAME
+  value: {{ include "invenio.postgresql.database" . }}
+- name: INVENIO_DB_PROTOCOL
+  value: "postgresql+psycopg2"
+- name: INVENIO_DB_PASSWORD
+{{- if or (and .Values.postgresql.enabled .Values.postgresql.auth.password) .Values.postgresqlExternal.password }}
+  value: {{ include "invenio.postgresql.password" .  | quote }}
+{{- else }}
+  valueFrom:
+    secretKeyRef:
+      name: {{ include "invenio.postgresql.secretName" .}}
+      key: {{ include "invenio.postgresql.secretKey" .}}
+{{- end }}
+- name: INVENIO_SQLALCHEMY_DATABASE_URI
+  value: "$(INVENIO_DB_PROTOCOL)://$(INVENIO_DB_USER):$(INVENIO_DB_PASSWORD)@$(INVENIO_DB_HOST):$(INVENIO_DB_PORT)/$(INVENIO_DB_NAME)"
```

Comment:
@lindhe What do you think about something like this? It should somehow address https://github.com/inveniosoftware/helm-invenio/issues/112

Reply:
Wow. Maybe. 👀 The logic with defining each part as a separate env is sound, no issues there. But composing a new env value based on other envs like this, I'm unfamiliar with. If it works as your code indicates, then yeah I think this looks very promising.

How does this work? Does Kubernetes support the `$(ENV)` syntax in string values, or does that happen inside the container? I feel lost as to why this works. 😅 I assumed you tried it and it seems to work?

Assuming it works, I really like this because I think it should be forwards-compatible with https://github.com/inveniosoftware/invenio-config/pull/57 and/or https://github.com/inveniosoftware/invenio-app-rdm/pull/2918 which is nice.

Reply:
I am not sure when it happens, in particular after reading this in the docs `Variables making use of others defined in the same context must come later in the list.`.
I would think the interpolation happens at run time inside the container. This `Environment` section of the output of `kubectl describe pod` 
```
     Environment:
      SECRET:   <set to the key 'secret' in secret 'compose-env'>  Optional: false
      PLAIN:    plain
      MESSAGE:  $(PLAIN) - $(SECRET)
```

Reply:
Sounds convincing! Wow, this is so powerful (and a bit confusing)! 😀 

I think this would be a great first step towards handling connection credentials better!

Reply:
As far as I can tell, this works now! 🚀